### ISSUE TITLE
bad flag value in twisted.conch.ssh.filetransfer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2902]
|<img alt="mmoskwa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mmoskwa reported|
|-|-|
|Trac ID|trac#2902|
|Type|defect|
|Created|2007-11-14 23:26:24Z|
|Branch|https://github.com/twisted/twisted/tree/acmodtime-sftp-2902|

flag FILEXFER_ATTR_ACMODTIME is set to 0x09 and it should be 0x08.  Calling setAttrs(path, attrs) when attrs contains keys for 'atime' and 'mtime' causes program to hang indefinitely.

Attachments:

* [blargh.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5a/f5a62fde9f8598f5c0f4afa74d2585e37ead65e2/b4c0fd782f79e56c1f563f7ddb1dee85622b2f62.diff) (54023 bytes) - added by mwh on 2008-10-29 12:35:01Z - 
* [blargh.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5a/f5a62fde9f8598f5c0f4afa74d2585e37ead65e2/cc59f97723bfcaad9d10f498e5b97976fe811d66.diff) (5743 bytes) - added by mwh on 2008-10-31 10:36:19Z - 
* [blargh.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f5a/f5a62fde9f8598f5c0f4afa74d2585e37ead65e2/4658cc373efa28d84536151b32a453d8735ed286.diff) (6536 bytes) - added by mwh on 2008-10-31 16:09:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2902 2902
type__defect defect
reporter__mmoskwa mmoskwa
priority__highest highest
milestone__ 
branch__branches_acmodtime_sftp_2902 branches/acmodtime-sftp-2902
branch_author__mwh__exarkun mwh, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1195082784000000 1195082784000000
changetime__1225553556000000 1225553556000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mmoskwa's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mmoskwa set owner to z3p|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

[#3512](https://github.com/twisted/twisted/issues/3512) was a duplicate of this

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson commented|
|-|-|

The attached patch fixes this, complete with rather ridiculous test.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_4"></a>|@mwhudson commented|
|-|-|

Oh, I should say that I know the test has style issues.  If the approach seems acceptable, I can clean those up.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @mwhudson|
|-|-|

I guess since only about 40 lines of the draft spec are being used, just those lines (that is, the #defines on page 5) should be included in the test with a link to the whole draft and a brief comment about why -02 is being used instead of a more recent draft.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

Like this?

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson commented|
|-|-|

And, in all seriousness, it is taking a ridiculous amount of effort to fix a one-character typo.  Something is broken in the Twisted world (maybe it's just the fact that conch should really be distributed separately from Twisted).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Sorry this is so hard.  If the goal is just to fix the issue ASAP, I would have tried for a more conventional testing approach.  I mentioned this on the channel, but maybe I should have included it in a comment on the ticket as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @mwhudson|
|-|-|

  1. The `sibpath` import added in this patch is no longer necessary, please remove it.
  1. `filexfer_spec_excerpts` and `test_constants_against_spec` are both coding-standard violations.  They should be `filexferSpecExcerpts` and `test_constantsAgainstSpec` respectively.
  1. Both the test case class and method need docstrings.  I suggest moving `filexferSpecExcerpts` an attribute of the test case, and moving the comment describing it into an `@ivar` in the class docstring.  The test method can simply say that it verifies the filetransfer constants against a pasted excerpt of the spec.

Given that none of this is functional feedback, please apply the changes to the patch and feel free to commit it.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @mwhudson<br>@glyph set status to `new`|
|-|-|

Now that you have commit, you can merge it yourself... :)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_15"></a>|@mwhudson set status to `closed`|
|-|-|

(In [25218]) Merge my patch: Fix the value of FILEXFER_ATTR_ACMODTIME in the sftp implementation.

Fix the value of FILEXFER_ATTR_ACMODTIME in twisted.conch.ssh.filetransfer, and
add tests that parse chunks of the relevant draft standard to make sure all the
constants are correct.

Author: mwh
Reviewer: glyph
Fixes: [#2902](https://github.com/twisted/twisted/issues/2902)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `reopened`|
|-|-|

(In [25219]) Revert 25218 - test suite regression

Reopens [#2902](https://github.com/twisted/twisted/issues/2902)

The newly added test fails on Python 2.3:

```
===============================================================================
[FAIL]: twisted.conch.test.test_filetransfer.TestConstants.test_constantsAgainstSpec

Traceback (most recent call last):
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/conch/test/test_filetransfer.py", line 676, in test_constantsAgainstSpec
    self.assertEqual(v, getattr(filetransfer, k))
twisted.trial.unittest.FailTest: not equal:
a = -2147483648
b = 2147483648L

-------------------------------------------------------------------------------
```

This appears to be due to the behavior of int("0x80000000", 0) on Python 2.3 (which
differs from the behavior on later versions of Python).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [25228]) Branching to 'acmodtime-sftp-2902'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [25230]) use int instead of long to ensure a proper interpretation of the unsigned 32 bit values in the spec

refs [#2902](https://github.com/twisted/twisted/issues/2902)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [25231]) use assertTrue instead of assert_ and add a message to make the test failure a bit more friendly

refs [#2902](https://github.com/twisted/twisted/issues/2902)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Building, results will be available at http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/acmodtime-sftp-2902

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Building again since that build had a bunch of spurious failures.  On the other hand that builder seems to have some chronic problems, so maybe it won't go away.  Results will be here:

http://buildbot.twistedmatrix.com/builders/winxp32-py2.4-select/builds/57

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set owner to @exarkun|
|-|-|

All right, that builder seems hosed for reasons unrelated to this branch.  Everything else checks out, though.

(Also, what button did you press to make the buildbot actually build that branch on all supported platforms?  I'd really like to do that more in the future.)

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `closed`|
|-|-|

(In [25243]) Merge acmodtime-sftp-2902

Author: mwh, exarkun
Reviewer: glyph
Fixes: [#2902](https://github.com/twisted/twisted/issues/2902)

Fix the value of FILEXFER_ATTR_ACMODTIME in the sftp implementation and add tests
which parse chunks of the relevant draft standard to make sure all the constants
are correct.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
